numpy/lib/arraysetops.py

```diff
@@ -254,42 +310,119 @@ def reshape_uniq(uniq):
         uniq = np.swapaxes(uniq, 0, axis)
         return uniq
 
+    consolidated_base = np.array_view_root(consolidated)
+    orig_base = np.array_view_root(orig_ar)
+    if return_index or return_inverse:
+        # We won't perform an in-place sort even if requested
+        mask_is_sorted = assume_sorted
+    elif orig_base is consolidated_base:
+        # We're operating on a view
+        mask_is_sorted = assume_sorted or sort_inplace
+    else:
+        # We're operating on a copy, so we might as well
+        # do an in-place sort on that copy
+        sort_inplace = True
+        mask_is_sorted = assume_sorted
+
     output = _unique1d(consolidated, return_index,
-                       return_inverse, return_counts)
-    output = (reshape_uniq(output[0]),) + output[1:]
+                       return_inverse, return_counts,
+                       return_mask, return_data, assume_sorted,
+                       sort_inplace, mask_is_sorted)
+
+    if return_data:
+        uniq = reshape_uniq(output[0])
+        output = (uniq,) + output[1:]
+
     return _unpack_tuple(output)
 
 
 def _unique1d(ar, return_index=False, return_inverse=False,
-              return_counts=False):
+              return_counts=False, return_mask=False,
+              return_data=True, assume_sorted=False,
+              sort_inplace=False, mask_is_sorted=False):
     """
     Find the unique elements of an array, ignoring shape.
     """
-    ar = np.asanyarray(ar).flatten()
+    ar = np.asanyarray(ar)
 
     optional_indices = return_index or return_inverse
-
+    optional_returns = (optional_indices or return_counts
+                        or return_mask or not return_data)
+
+    if (not optional_indices and not assume_sorted
+        and return_mask and not mask_is_sorted):
+        # Will need the permutation to compute masks
+        optional_indices = True
+    if ((not optional_indices and not assume_sorted and not sort_inplace)
+        or ar.ndim != 1):
+        # Otherwise, we don't need to make a copy
+        if sort_inplace and ar.ndim == 2 and ar.shape[1] == 1:
+            # consolidated (n x 1) matrix, no need to copy
+            ar = ar.ravel()
+        else:
+            ar = ar.flatten()
+            if mask_is_sorted and not assume_sorted:
+                # We'll work on a copy, so mask cannot be sorted
+                mask_is_sorted = False
+                if return_mask:
+                    optional_indices = True
+
+    if ar.size == 0:
+        ret = ()
+        if return_data:
+            ret = (ar,)
+        if return_index:
+            ret += (np.empty(0, np.intp),)
+        if return_inverse:
+            ret += (np.empty(0, np.intp),)
+        if return_counts:
+            ret += (np.empty(0, np.intp),)
+        if return_mask:
+            ret += (np.empty(0, np.bool_),)
+        return ret
```

Comment:
I don't think this is needed either

Reply:
Yep, sorry, I botched the merge. Fixed now.